File: ags.go
// Package ags parses, validates, identifies, and plans Agentic Graph
// Specification 1.0 documents.
package ags

const (
	AGSVersion     = "1.0"
	SupportVersion = "1.0.2"
)

type Document map[string]any

type Finding struct {
	Code     string `json:"code"`
	Severity string `json:"severity"`
	Message  string `json:"message"`
	Pointer  string `json:"pointer,omitempty"`
}

type Report struct {
	Document Document  `json:"document,omitempty"`
	Findings []Finding `json:"findings"`
	Errors   []Finding `json:"errors"`
	Warnings []Finding `json:"warnings"`
	OK       bool      `json:"ok"`
}

func (r *Report) add(code, severity, message, pointer string) {
	f := Finding{Code: code, Severity: severity, Message: message, Pointer: pointer}
	r.Findings = append(r.Findings, f)
	if severity == "error" {
		r.Errors = append(r.Errors, f)
	} else {
		r.Warnings = append(r.Warnings, f)
	}
}

type EffectiveEdge struct {
	From string `json:"from"`
	To   string `json:"to"`
	Kind string `json:"kind"`
	When string `json:"when,omitempty"`
}

type Plan struct {
	GraphID                 string          `json:"graph_id"`
	GraphDigest             string          `json:"graph_digest"`
	Order                   []string        `json:"order"`
	Entrypoints             []string        `json:"entrypoints"`
	EffectiveEdges          []EffectiveEdge `json:"effective_edges"`
	Reachable               []string        `json:"reachable"`
	Unreachable             []string        `json:"unreachable"`
	TierHistogram           map[string]int  `json:"tier_histogram"`
	WorstCaseNodeExecutions int             `json:"worst_case_node_executions"`
	Executable              bool            `json:"executable"`
	UnsupportedFeatures     []string        `json:"unsupported_features"`
}

File: java/src/main/java/io/github/alexmercedcoder/ags/Ags.java
package io.github.alexmercedcoder.ags;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.util.List;
import java.util.Map;

/** Shared AGS constants and immutable result types. */
public final class Ags {
  /** AGS specification version implemented by this library. */
  public static final String SPEC_VERSION = "1.0";
  /** Java support-library version. */
  public static final String SUPPORT_VERSION = "1.0.3";

  private Ags() {}

  /** One machine-readable validation diagnostic. */
  public record Finding(String code, String severity, String message, String pointer) {}

  /** Complete parsing and validation result. */
  public record ValidationReport(
      ObjectNode document,
      List<Finding> findings,
      List<Finding> errors,
      List<Finding> warnings,
      boolean ok) {}

  /** Normalized dependency or explicitly declared graph edge. */
  public record EffectiveEdge(String from, String to, String kind, String when) {}

  /** Deterministic, non-executing conformance-level-0 graph plan. */
  public record GraphPlan(
      String graphId,
      String graphDigest,
      List<String> order,
      List<String> entrypoints,
      List<EffectiveEdge> effectiveEdges,
      List<String> reachable,
      List<String> unreachable,
      Map<String, Integer> tierHistogram,
      long worstCaseNodeExecutions,
      boolean executable,
      List<String> unsupportedFeatures) {}

  static ObjectNode object(JsonNode node) {
    return node != null && node.isObject() ? (ObjectNode) node : JsonSupport.emptyObject();
  }
}
